packages/logging/src/metadata.js

```diff
@@ -0,0 +1,297 @@
+/*!
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*!
+ * @module logging/metadata
+ */
+
+'use strict';
+var is = require('is');
+
+/**
+ * The abstract events emitted to invokers of getDefaultResource and
+ * getProjectId which are fired upon evaluation of the response from the
+ * metadata service.
+ * @const REQUEST_EVENTS
+ */
+var REQUEST_EVENTS = {
+  SUCCESS: 'METADATA_REQUEST_SUCCESS',
+  FAILURE: 'METADATA_REQUEST_FAILURE'
+};
+
+/**
+ * The Metadata class attempts to contact the metadata service and determine,
+ * based on request success and environment variables, what type of resource
+ * the library is operating on.
+ *
+ * @constructor
+ * @alias module:metadata
+ *
+ * @resource [Logs Resource API Documentation]{@link https://cloud.google.com/logging/docs/api/reference/rest/v2/MonitoredResource}
+ *
+ * @param {Duplex} requestStream - A stream returned by a request style
+ *  invocation to the metadata project id endpoint
+ */
+function Metadata(requestStream) {
+  this.metadataProjectId_ = null;
+  this.metadataContactError_ = null;
+  this.metadataRequestStream_ = requestStream;
+  this.attachRequestStreamListeners_(requestStream);
+}
+
+/**
+ * Attaches core request lifecycle event listeners on the request stream
+ * provided to the constuctor.
+ *
+ * @private
+ */
+Metadata.prototype.attachRequestStreamListeners_ = function(stream) {
+  var self = this;
```

Comment:
There's actually a major simplification we can make, which is using the auth client to get the project ID. It detects it from all of the same places you are here.

You could cache the Logging instance in the Metadata constructor, and call `logging.authClient.getProjectId(function(err, projectId) {})`.

```js
function Metadata(logging) {
  this.logging = logging
}

Metadata.prototype.getProjectId = function(callback) {
  if (this.logging.projectId) {
    setImmediate(function() { callback(null, projectId) })
    return
  }

  this.logging.getProjectId(callback)
}
```

> Given the case where we have multiple callers I still have to have a way of queuing callbacks on the resolution

How do we end up in that situation, using a code call like this:

```js
log.write([entries], function(err) {})
```

> so no user-space queuing

I'm still opposed to running phantom requests in the background. It should run at the time the user has a request to send.